src/backend/distributed/executor/multi_client_executor.c

```diff
@@ -94,6 +94,13 @@ MultiClientConnect(const char *nodeName, uint32 nodePort, const char *nodeDataba
 	char *effectiveDatabaseName = NULL;
 	char *effectiveUserName = NULL;
 
+	if (IsModifyingTransaction)
+	{
+		ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+						errmsg("cannot open new connection during modifying "
```

Comment:
I'm not entirely sure how, but it seems this message could be made a bit easier to understand for users.  Maybe something like "cannot open new connections in a transaction that has performed modifications", hint: After a modification has been executed only router-executable queries to exactly those shards are allowed. or something.